Cabal/src/Distribution/Simple/GHC.hs

```diff
@@ -996,6 +1002,45 @@ startInterpreter verbosity progdb comp platform packageDBs = do
   (ghcProg, _) <- requireProgram verbosity ghcProgram progdb
   runGHC verbosity ghcProg comp platform replOpts
 
+
+runReplOrWriteFlags
+  :: Verbosity
+  -> ConfiguredProgram
+  -> Compiler
+  -> Platform
+  -> ReplOptions
+  -> GhcOptions
+  -> BuildInfo
+  -> ComponentLocalBuildInfo
+  -> PackageName
+  -> IO ()
+runReplOrWriteFlags verbosity ghcProg comp platform rflags replOpts bi clbi pkg_name =
+  case replOptionsFlagOutput rflags of
+    NoFlag -> runGHC verbosity ghcProg comp platform replOpts
+    Flag out_dir -> do
+      src_dir <- getCurrentDirectory
```

Comment:
It's the folder where the `.cabal` file is. And this is find because there is already the assumption that `Setup.hs` is executed in the folder where the .cabal file is. (This is why all the error messages you get from cabal have paths relative to the .cabal file rather than the cabal.project root.)

See `internalSetupMethod` in `cabal-install/SetupWrapper.hs` which changes the directory.